crates/miden-objects/src/asset/vault.rs

```diff
@@ -261,11 +261,16 @@ impl AssetVault {
 
 impl Serializable for AssetVault {
     fn write_into<W: ByteWriter>(&self, target: &mut W) {
-        // TODO: determine total number of assets in the vault without allocating the vector
-        let assets = self.assets().collect::<Vec<_>>();
+        // Get the number of assets directly without iterating
+        let count = self.asset_tree.num_entries();
 
-        target.write_usize(assets.len());
-        target.write_many(&assets);
+        // Write the count
+        target.write_usize(count);
+
+        // Write each asset directly from the iterator
+        for entry in self.asset_tree.entries() {
+            Asset::new_unchecked(entry.1).write_into(target);
+        }
```

Comment:
```suggestion
        for (_asset_key, asset) in self.asset_tree.entries() {
            Asset::new_unchecked(*asset).write_into(target);
        }
```

Nit: I would write it like this.